File: leetcode/python/Remove Duplicates from Sorted List/Remove_Duplicates_from_Sorted_List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a ListNode
    # @return a ListNode
    def deleteDuplicates(self, head):
        if not head:
            return head
        h = head
        pre = head
        while h.next:
            h = h.next
            if pre.val != h.val:
                pre.next = h
                pre = pre.next
        pre.next = None
        return head

from ListNode import ListNode
solu = Solution()
node1 = ListNode(1)
node2 = ListNode(1)
node3 = ListNode(2)
node4 = ListNode(4)
node5 = ListNode(5)
node1.next = node2
node2.next = node3
#node3.next = node4
#node4.next = node5
head = solu.deleteDuplicates(node1,)
while head:
    print head.val
    head = head.next

File: leetcode/cpp/RemoveDuplicatesfromSortedList.cpp
//
// Created by zz on 2017/10/31.
// Time:  O(n)
// Space: O(1)

struct ListNode {
    int val;
    ListNode *next;

    ListNode(int x) : val(x), next(nullptr) {}
};


class Solution {
public:
    ListNode *deleteDuplicates(ListNode *head) {
        ListNode *curr = head;
        while (curr) {
            ListNode *runner = curr;
            while (runner && curr->val == runner->val) {
                runner = runner->next;
            }
            curr->next = runner;
            curr = runner;
        }
        return head;
    }
};